File: javascript/benchmark.js
const fs = require('fs')

if (process.argv.length !== 3) {
  console.log('Usage: node benchmark.js <filename>')
  process.exit(1)
}

function measure(data, pattern) {
  const start = process.hrtime()

  const regex = new RegExp(pattern, 'g')
  const matches = data.match(regex)

  const end = process.hrtime(start);
  
  console.log((end[0] * 1e9 + end[1]) / 1e6 + ' - ' + matches.length);
}

const data = fs.readFileSync(process.argv[2], 'utf8')

// Email
measure(data, '[\\w.+-]+@[\\w.-]+\\.[\\w.-]+')

// URI
measure(data, '[\\w]+:\\/\\/[^\\/\\s?#]+[^\\s?#]+(?:\\?[^\\s#]*)?(?:#[^\\s]*)?')

// IP
measure(data, '(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9])\\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9])')

File: dart/benchmark.dart
import 'dart:io';

main(List<String> arguments) {
  if (arguments.length != 1) {
    print('Usage: dart bencharmark.dart <filename>');
    exit(1);
  }

  new File(arguments[0])
    .readAsString()
    .then((String data) {
      // Email
      measure(data, r"[\w.+-]+@[\w.-]+\.[\w.-]+");

      // URI
      measure(data, r"[\w]+:\/\/[^\/\s?#]+[^\s?#]+(?:\?[^\s#]*)?(?:#[^\s]*)?");

      // IP
      measure(data, r"(?:(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9])\.){3}(?:25[0-5]|2[0-4][0-9]|[01]?[0-9][0-9])");
    });
}

measure(data, pattern){
  var stopwatch = new Stopwatch()..start();

  RegExp exp = new RegExp(pattern);
  Iterable<Match> matches = exp.allMatches(data);
  var count = matches.length;

  stopwatch.stop();

  print('${stopwatch.elapsedMicroseconds / 1e3} ${count}');
}
